File: frontend/src/lib/billing.ts
import { apiRequest } from "@/lib/api-client";

export type BillingCycle = "monthly" | "annual";
export type PaymentMethod = "card" | "pix_automatic";

export type CheckoutSubscribePayload = {
  billing_cycle: BillingCycle;
  payment_method: PaymentMethod;
  cpf: string;
  card_holder_name?: string;
  card_number?: string;
  card_expiry_month?: string;
  card_expiry_year?: string;
  card_cvv?: string;
  holder_postal_code?: string;
  holder_address_number?: string;
  holder_phone?: string;
};

export type CheckoutSubscribeResponse = {
  status: "pending" | "confirmed";
  payment_method: PaymentMethod;
  external_subscription_id: string;
  amount: number;
  currency: string;
  billing_cycle: BillingCycle;
  message: string;
  pix_qr_code?: string | null;
  pix_copy_paste?: string | null;
  mock_checkout: boolean;
};

export type CheckoutStatus = {
  has_pending_checkout: boolean;
  checkout_status?: "pending" | "authorized" | "cancelled" | "failed" | null;
  payment_status?: "pending" | "confirmed" | "overdue" | "canceled" | "processing" | null;
  payment_method?: PaymentMethod | null;
  billing_cycle?: BillingCycle | null;
  amount?: number | null;
  currency?: string;
  external_subscription_id?: string | null;
  pix_qr_code?: string | null;
  pix_copy_paste?: string | null;
  checkout_created_at?: string | null;
  message?: string | null;
};

export type BillingSubscription = {
  plan_id: string;
  subscription_status: string;
  subscription_started_at: string | null;
  subscription_ends_at: string | null;
  subscription_renews_at: string | null;
  billing_cycle: BillingCycle | null;
  subscription_source: string;
  payment_method: string | null;
  can_manage_billing: boolean;
  manage_url: string | null;
};

export type BillingCheckoutHistoryItem = {
  id?: number | null;
  billing_cycle?: BillingCycle | null;
  payment_method?: string | null;
  status?: string | null;
  external_subscription_id?: string | null;
  amount?: number | null;
  currency?: string;
  created_at?: string | null;
  updated_at?: string | null;
  can_resume?: boolean;
  can_retry?: boolean;
};

export type BillingEventHistoryItem = {
  id?: number | null;
  event_type?: string | null;
  event_key?: string | null;
  processed_at?: string | null;
  payment_status?: string | null;
};

export type BillingHistorySubscription = {
  plan_id?: string | null;
  status?: string | null;
  started_at?: string | null;
  ends_at?: string | null;
  renews_at?: string | null;
  billing_cycle?: BillingCycle | null;
  subscription_source?: string | null;
  payment_method?: string | null;
  external_subscription_id?: string | null;
};

export type BillingHistory = {
  subscription: BillingHistorySubscription | null;
  checkouts: BillingCheckoutHistoryItem[];
  events: BillingEventHistoryItem[];
  pending_checkout?: BillingCheckoutHistoryItem | null;
};

export async function subscribeCheckout(
  accessToken: string,
  payload: CheckoutSubscribePayload,
) {
  return apiRequest<CheckoutSubscribeResponse>("/api/v1/billing/checkout/subscribe", {
    accessToken,
    method: "POST",
    body: payload,
  });
}

export async function loadCheckoutStatus(accessToken: string) {
  return apiRequest<CheckoutStatus>("/api/v1/billing/checkout/status", { accessToken });
}

export async function resumePendingCheckout(
  accessToken: string,
  externalSubscriptionId: string,
) {
  return apiRequest<CheckoutStatus>("/api/v1/billing/checkout/resume", {
    accessToken,
    method: "POST",
    body: { external_subscription_id: externalSubscriptionId },
  });
}

export async function abandonPendingCheckout(
  accessToken: string,
  externalSubscriptionId?: string | null,
) {
  return apiRequest<CheckoutStatus>("/api/v1/billing/checkout/abandon", {
    accessToken,
    method: "POST",
    body: {
      external_subscription_id: externalSubscriptionId || undefined,
    },
  });
}

export async function refreshBillingSubscription(accessToken: string) {
  return apiRequest<{
    updated: boolean;
    plan_id?: string;
    subscription_status?: string;
    reason?: string;
  }>("/api/v1/billing/refresh", {
    accessToken,
    method: "POST",
  });
}

export async function loadBillingSubscription(accessToken: string) {
  return apiRequest<BillingSubscription>("/api/v1/billing/subscription", { accessToken });
}

export async function loadBillingHistory(accessToken: string) {
  return apiRequest<BillingHistory>("/api/v1/billing/history", { accessToken });
}

export async function cancelBillingSubscription(accessToken: string) {
  return apiRequest<{
    subscription_status: string;
    subscription_ends_at: string | null;
    external_subscription_id: string;
  }>("/api/v1/billing/subscription/cancel", {
    accessToken,
    method: "POST",
  });
}

File: backend/app/schemas/billing.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

BillingCycle = Literal["monthly", "annual"]
PaymentMethod = Literal["card", "pix_automatic"]


class CheckoutSubscribeRequest(BaseModel):
    billing_cycle: BillingCycle
    payment_method: PaymentMethod
    cpf: str = Field(min_length=11, max_length=14)
    card_holder_name: str | None = Field(default=None, max_length=120)
    card_number: str | None = Field(default=None, max_length=19)
    card_expiry_month: str | None = Field(default=None, max_length=2)
    card_expiry_year: str | None = Field(default=None, max_length=4)
    card_cvv: str | None = Field(default=None, max_length=4)
    holder_postal_code: str | None = Field(default=None, max_length=9)
    holder_address_number: str | None = Field(default=None, max_length=20)
    holder_phone: str | None = Field(default=None, max_length=20)


class CheckoutSubscribeResponse(BaseModel):
    status: Literal["pending", "confirmed"]
    payment_method: PaymentMethod
    external_subscription_id: str
    amount: float
    currency: str = "BRL"
    billing_cycle: BillingCycle
    message: str
    pix_qr_code: str | None = None
    pix_copy_paste: str | None = None
    mock_checkout: bool = False


class CancelSubscriptionResponse(BaseModel):
    subscription_status: str
    subscription_ends_at: datetime | None = None
    external_subscription_id: str


class BillingSubscriptionView(BaseModel):
    plan_id: str
    subscription_status: str
    subscription_started_at: datetime | None = None
    subscription_ends_at: datetime | None = None
    subscription_renews_at: datetime | None = None
    billing_cycle: str | None = None
    subscription_source: str = "system"
    payment_method: str | None = None
    can_manage_billing: bool = False
    manage_url: str | None = None


class BillingRefreshResponse(BaseModel):
    updated: bool
    plan_id: str | None = None
    subscription_status: str | None = None
    reason: str | None = None


class CheckoutStatusResponse(BaseModel):
    has_pending_checkout: bool
    checkout_status: Literal["pending", "authorized", "cancelled", "failed"] | None = None
    payment_status: Literal["pending", "confirmed", "overdue", "canceled", "processing"] | None = (
        None
    )
    payment_method: PaymentMethod | None = None
    billing_cycle: BillingCycle | None = None
    amount: float | None = None
    currency: str = "BRL"
    external_subscription_id: str | None = None
    pix_qr_code: str | None = None
    pix_copy_paste: str | None = None
    checkout_created_at: datetime | None = None
    message: str | None = None


class BillingCheckoutHistoryItem(BaseModel):
    id: int | None = None
    billing_cycle: BillingCycle | None = None
    payment_method: str | None = None
    status: str | None = None
    external_subscription_id: str | None = None
    amount: float | None = None
    currency: str = "BRL"
    created_at: datetime | None = None
    updated_at: datetime | None = None
    can_resume: bool = False
    can_retry: bool = False


class ResumeCheckoutRequest(BaseModel):
    external_subscription_id: str = Field(min_length=1, max_length=120)


class AbandonCheckoutRequest(BaseModel):
    external_subscription_id: str | None = Field(default=None, max_length=120)


class BillingEventHistoryItem(BaseModel):
    id: int | None = None
    event_type: str | None = None
    event_key: str | None = None
    processed_at: datetime | None = None
    payment_status: str | None = None


class BillingHistoryResponse(BaseModel):
    subscription: dict[str, Any] | None = None
    checkouts: list[BillingCheckoutHistoryItem] = Field(default_factory=list)
    events: list[BillingEventHistoryItem] = Field(default_factory=list)
    pending_checkout: BillingCheckoutHistoryItem | None = None
